### PR TITLE
pkg/lightning: only retry unfinished ranges when retry import engine

### PR DESCRIPTION
**Pull request by glorv**
### What problem does this PR solve? 
In our multi-lightning tests, sometimes the import index engine phase encounters error retry after running for hours. In this condition, a lot of ranges have already finished ingesting, so we can avoid retrying these finished ranges in local backend when retry importing engine to make import retry faster.

### What is changed and how it works?
- Save finished ranges into `LocalEngine` after ingest succeeded. And always calculate the retry ranged based on current finished ranges.

TODO:
Shall we also save the finished ranges info into pebble like the `engineMeta` as a checkpoint?

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

Side effects

Related changes

 - Need to cherry-pick to the release branch

### Release Note

 - Only retry unfinished ranges when retry import full engine.

### CODE REVIEW

**Comment by glorv** [comment]
@kennytm @Little-Wallace PTAL

File: pkg/lightning/backend/local/local.go [hunk 1]
```diff
@@ -902,11 +903,19 @@ func splitRangeBySizeProps(fullRange Range, sizeProps *sizeProperties, sizeLimit
...
+		}
 		curSize += p.Size
 		curKeys += p.Keys
 		if int64(curSize) >= sizeLimit || int64(curKeys) >= keysLimit {
-			ranges = append(ranges, Range{start: curKey, end: p.Key})
-			curKey = p.Key
+			// in case the sizeLimit or keysLimit is too small
+			endKey := p.Key
+			if bytes.Equal(curKey, endKey) {
+				endKey = nextKey(endKey)
```
Comment (lance6716) [inline pkg/lightning/backend/local/local.go:915]: could we just return

File: pkg/lightning/backend/local/local.go [hunk 2]
```diff
@@ -1309,18 +1326,18 @@ func (local *local) writeAndIngestByRanges(ctx context.Context, engineFile *File
...
 
-func (r *syncdRanges) add(g Range) {
+func (r *syncedRanges) add(g Range) {
 	r.Lock()
 	r.ranges = append(r.ranges, g)
 	r.Unlock()
 }
 
-func (r *syncdRanges) take() []Range {
+func (r *syncedRanges) take() []Range {
```
Comment (lance6716) [inline pkg/lightning/backend/local/local.go:1340]: not used now

File: pkg/lightning/backend/local/local.go [hunk 3]
```diff
@@ -1380,25 +1405,95 @@ func (local *local) ImportEngine(ctx context.Context, engineUUID uuid.UUID) erro
...
+func (engine *File) unfinishedRanges(ranges []Range) []Range {
+	engine.finishedRanges.Lock()
+	defer engine.finishedRanges.Unlock()
+
+	engine.finishedRanges.ranges = dedupAndMergeRanges(engine.finishedRanges.ranges)
+
+	return filterOverlapRange(ranges, engine.finishedRanges.ranges)
+}
+
+func dedupAndMergeRanges(ranges []Range) []Range {
```
Comment (lance6716) [inline pkg/lightning/backend/local/local.go:1429]: can it deduplicate, and handle one range totally covers another?

Comment (glorv) [reply pkg/lightning/backend/local/local.go:1429]: Yes，If several range overlap with each other, the result will be the smallest start and the largest end

Comment (lance6716) [reply pkg/lightning/backend/local/local.go:1429]: the "dedup" makes me think of "[[1,2], [1,2]] -> [[1,2]]" 😂

File: pkg/lightning/backend/local/local.go (hunk 1 above)
Comment (glorv) [reply pkg/lightning/backend/local/local.go:915]: If the region-split-size is set to a very small value that one kv can exceed this threshold, then the logic can exact cover this threshold, though such small threshold threshold is only used in tests

File: pkg/lightning/backend/local/local.go (hunk 3 above)
Comment (glorv) [reply pkg/lightning/backend/local/local.go:1429]: The `TestDedupAndMergeRanges` unit test has test cases that covers this condition

Comment (glorv) [reply pkg/lightning/backend/local/local.go]: OK, I rename the function name to `sortAndMergeRanges`😅

File: pkg/lightning/backend/local/local.go [hunk 4]
```diff
@@ -854,19 +855,19 @@ func (local *local) WriteToTiKV(
 		zap.Int64("buf_size", bytesBuf.totalSize()),
 		zap.Stringer("takeTime", time.Since(begin)))
 
-	var remainRange *Range
+	finishedRange := regionRange
 	if iter.Valid() && iter.Next() {
 		firstKey := append([]byte{}, iter.Key()...)
-		remainRange = &Range{start: firstKey, end: regionRange.end}
+		finishedRange = Range{start: regionRange.start, end: firstKey}
```
Comment (Little-Wallace) [inline pkg/lightning/backend/local/local.go:861]: Shall it be `regionRange` or `start` passed by method parameter?

Comment (glorv) [reply pkg/lightning/backend/local/local.go:861]: The `start`, `end` passed from method parameter is the Range calculated by lightning, and it's range may covers multiple regions. The`regionRange` is the actual range the intersect  target region and [start, end). So the finishedRange's start should be the regionRange's start

File: pkg/lightning/backend/local/local.go (hunk 1 above)
Comment (lance6716) [reply pkg/lightning/backend/local/local.go:915]: I see, `return` will cause the range merged into next one.

**Review by lance6716 — APPROVED** [review]

**Pull request merged**